src/main/java/io/cdap/plugin/batch/source/MongoDBBatchSource.java

```diff
@@ -0,0 +1,253 @@
+    conf.clear();
+
+    MongoConfigUtil.setInputFormat(conf, MongoInputFormat.class);
+    MongoConfigUtil.setInputURI(conf, config.getConnectionString());
+    MongoConfigUtil.setFields(conf, config.getProjectionDocument());
+    if (!Strings.isNullOrEmpty(config.inputQuery)) {
+      MongoConfigUtil.setQuery(conf, config.inputQuery);
+    }
+    if (!Strings.isNullOrEmpty(config.authConnectionString)) {
+      MongoConfigUtil.setAuthURI(conf, config.authConnectionString);
+    }
+
+    emitLineage(context);
+    context.setInput(Input.of(config.referenceName,
+                              new SourceInputFormatProvider(MongoConfigUtil.getInputFormat(conf), conf)));
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    bsonObjectToRecordTransformer = new BSONObjectToRecordTransformer(config.getSchema());
+  }
+
+  @Override
+  public void transform(KeyValue<Object, BSONObject> input, Emitter<StructuredRecord> emitter) throws Exception {
+    try {
+    BSONObject bsonObject = input.getValue();
+    emitter.emit(bsonObjectToRecordTransformer.transform(bsonObject));
+    } catch (Exception e) {
+      switch (config.getErrorHandling()) {
+        case SKIP:
+          LOG.warn("Failed to process record, skipping it", e);
+          break;
+        case FAIL_PIPELINE:
+          throw new RuntimeException("Failed to process record", e);
+        default:
+          // this should never happen because it is validated at configure and prepare time
+          throw new IllegalStateException(String.format("Unknown error handling strategy '%s'",
+                                                        config.getErrorHandling()));
+      }
+    }
+  }
+
+  private void emitLineage(BatchSourceContext context) {
+    LineageRecorder lineageRecorder = new LineageRecorder(context, config.referenceName);
+    lineageRecorder.createExternalDataset(config.getSchema());
+    List<Schema.Field> fields = Objects.requireNonNull(config.getSchema()).getFields();
+    if (fields != null && !fields.isEmpty()) {
+      lineageRecorder.recordRead("Read",
+                                 String.format("Read from '%s' MongoDB collection.", config.collection),
+                                 fields.stream().map(Schema.Field::getName).collect(Collectors.toList()));
+    }
+  }
+
+  /**
+   * Config class for {@link MongoDBBatchSource}.
+   */
+  public static class MongoDBSourceConfig extends MongoDBConfig {
+
+    public static final Set<Schema.Type> SUPPORTED_SIMPLE_TYPES = ImmutableSet.of(Schema.Type.BOOLEAN, Schema.Type.INT,
+                                                                                  Schema.Type.DOUBLE, Schema.Type.BYTES,
+                                                                                  Schema.Type.LONG, Schema.Type.STRING,
+                                                                                  Schema.Type.ARRAY, Schema.Type.RECORD,
```

Comment:
same question about record, array, map validation as for Sink

Reply:
Changed to use [MongoDBUtil#validateSchema](https://github.com/vladhlinskiy/mongodb-plugins/pull/1/files#diff-55d7876dc5d84b30dfba3eb7c10d9b85R47) and wrap validation exceptions into corressponding `InvalidConfigPropertyException`:

```
        try {
          MongoDBUtil.validateSchema(parsedSchema, SUPPORTED_LOGICAL_TYPES, SUPPORTED_SIMPLE_TYPES);
        } catch (IllegalArgumentException e) {
          throw new InvalidConfigPropertyException(e.getMessage(), e, MongoDBConstants.SCHEMA);
        }
```